prov/sockets/src/sock_progress.c

```diff
@@ -2629,10 +2630,12 @@ static void sock_pe_poll(struct sock_pe *pe)
 	if (pe->waittime && ((fi_gettime_ms() - pe->waittime) < sock_pe_waittime))
 		return;
 
-	if (dlist_empty(&pe->tx_list) && dlist_empty(&pe->rx_list))
+	pthread_mutex_lock(&pe->list_lock);
+	if (dlist_empty(&pe->tx_list) && dlist_empty(&pe->rx_list)) {
+		pthread_mutex_unlock(&pe->list_lock);
 		return;
+	}
 
-	pthread_mutex_lock(&pe->list_lock);
 	if (!dlist_empty(&pe->tx_list)) {
```

Comment:
But I think it's checking if *both* lists are empty.  Without a lock, you can't check both atomically.  That's why I moved the lock up (in addition to the helgrind complaint).

My $0.02: Performance is not the issue here; I think code clarity is more important.  I.e., even though `dlist_empty()` is nominally doing just a pointer check, we might as well strive to have clear code that is helgrind-clean (i.e,. so we don't have to examine each helgrind output and re-remind ourselves, "Oh yah, that one is (probably) ok."